docs/manual/resource-leak-checker.tex

```diff
@@ -0,0 +1,333 @@
+\htmlhr
+\chapterAndLabel{Resource Leak Checker for must-call obligations}{resource-leak-checker}
+
+The Resource Leak Checker guarantees that the program fulfills every object's
+must-call obligations before it is de-allocated.
+
+A resource leak occurs when a program does not explicitly dispose of some finite
+resource, such as a socket, file descriptor, or database connection.  To dispose
+of the resource, the program should call some method on an object.
+(De-allocating or garbage-collecting the object is not sufficient.)  For
+example, the program must call \<close()> on every object that implements the
+interface \<java.io.Closeable>.
+
+The Resource Leak Checker can check any property of the form ``the programmer
+must call all methods \emph{M} on object \emph{O} before \emph{O} is
+de-allocated''.  For resource leaks, by default \emph{M} is the set containing
+\<close()> and \emph{O} is any object that implements the interface
+\<java.io.Closeable>.  You can extend this guarantee to other types and methods
+by writing \<@MustCall> or \<@InheritableMustCall> annotations, as described in
+Section~\ref{must-call-annotations}.
+
+The Resource Leak Checker works in three stages:
+\begin{enumerate}
+\item The Must Call Checker (\chapterpageref{must-call-checker})
+  over-approximates each expression's must-call methods as a
+  \refqualclass{checker/mustcall/qual}{MustCall} type.
+\item The Called Methods Checker (\chapterpageref{called-methods-checker})
+  under-approximates each expression's definitely-called methods as a
+  \refqualclass{checker/calledmethods/qual}{CalledMethods} type.
+\item When any program element goes out of scope (i.e., it is ready to be
+  de-allocated), the Resource Leak Checker compares the types
+  \<@MustCall(\emph{MC})> and \<@CalledMethods(\emph{CM})>.  It reports an error
+  if there exists some method in \emph{MC} that is not in \emph{CM}.
+\end{enumerate}
+
+
+\sectionAndLabel{How to run the Resource Leak Checker}{resource-leak-run-checker}
+
+\begin{Verbatim}
+javac -processor ResourceLeak MyFile.java ...
+\end{Verbatim}
+
+The Resource Leak Checker supports the following optional command-line arguments,
+in addition to those listed in Section~\ref{called-methods-run-checker} for
+the Called Methods Checker:
+\begin{itemize}
+\item The \<-AnoLightweightOwnership> option disables ownership annotations,
```

Comment:
The manual section should say why a user would supply these command-line arguments.  I think it's just for efficiency, in cases where the program does not use a particular feature.

How big a difference does each one make in practice?  We should have data about that.  We might or might not want to put that in the section.  More importantly, it will help us decide whether the options are worth retaining, or whether we should simplify the implementation and the documentation by eliminating them.

Reply:
These options exist to allow us to run an ablation study to figure out how much each contributes to filtering false positives. None of them provide any efficiency advantage that we've measured, though there might be a small one. The data that we do have indicates that the first two of these options (`-AnoLightweightOwnership` and `-AnoResourceAliases`) are critical to the checker's false positive rate (177 and 316 false positives prevented, respectively, in the three case studies from the paper, which had 116 other false positives). The third option is also only negative (i.e. add more false positives), but less so than the other two with only 13 FPs prevented.

All this is to say that these options are only useful for an experimenter interested in whether the various features of the checker are useful. I'm doubt most users will care about that, so maybe we should put these in their own little section that says so. But, I do think they're worth keeping around.

Reply:
Since all 4 of these options are of no interest to users, I strongly recommend not documenting them in the manual.  I'm OK with leaving them in the code.